docs:#465 deploy for development

micheleRP opened the pull request:
resolves #465 

This adds a Deploy for Development page (similar to Deploy for Production, but not setting production mode or redpanda-tuner). 

preview: 
- https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment/
- https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/
- https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/

larsenpanda commented:
The reason someone would want to deploy for development is actually the majority of people getting onboarded with Redpanda until the final stage which is deploying to production, all the environments leading up to that are non-production. I think the only difference however would be whether they run rpk mode production or rpk mode development. The latter allows for auto topic create I believe,, i'm not totally clear on what development mode means in terms of configurations in that "mode" but we should use this opportunity to document that.

andrwng commented:
[Here](https://github.com/redpanda-data/redpanda/blob/740141f25aecd783598ea351aa0f42d268670925/src/go/rpk/pkg/cli/cmd/redpanda/start.go#L1042) the are the configs `rpk` sets when using developer mode:
```
	cfg.Redpanda.Other["auto_create_topics_enabled"] = true
	cfg.Redpanda.Other["group_topic_partitions"] = 3
	cfg.Redpanda.Other["storage_min_free_bytes"] = 10485760
	cfg.Redpanda.Other["topic_partitions_per_shard"] = 1000
	cfg.Redpanda.Other["fetch_reads_debounce_timeout"] = 10
	cfg.Redpanda.Other["group_initial_rebalance_delay"] = 0
	cfg.Redpanda.Other["log_segment_size_min"] = 1
```

and [here](https://docs.redpanda.com/docs/reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode) are docs for other things it does.

Generally the theme of these configs seems to be for setting up an underpowered/underprovisioned environment. E.g. whereas by default `group_topic_partitions` is 16, it's 3 here; whereas the default `storage_min_free_bytes` is 5G, it's 1G here. Auto-create is enabled: that's more for convenience in testing AFAICT (it's got nothing to do with the size of the deployment).

>Deploy for Development instead of just running our quickstart

One way to think about dev clusters is as a sort of staging area, everything else outside of Redpanda might be configured in the way they expect in prod (networking, permissions, etc), but the cluster as a whole might be smaller and it matters less if the data is accidentally deleted.

>I think the only difference however would be whether they run rpk mode production or rpk mode development

That said, I'm not sure developer mode in RPK really aligns with the expectation here. Setting underpowered configs is fine, but I'm less confident that we want to not fsync in a dev cluster, e.g. if that environment is used for chaos testing, day-2 ops, etc.

Maybe @r-vasquez can chime in with what the history behind developer mode is, since I see he added the `--mode` flag.

micheleRP commented:
Thank you, @andrwng! I added some context from you to the intro. I also added the prereqs (except memory) that we just added in the [Deploy for Production page](https://docs.redpanda.com/docs/deploy/deployment-option/self-hosted/manual/production/production-deployment/). Can you or @r-vasquez please confirm if we should include these prereqs this Deploy for Development page?

r-vasquez commented:
The history behind this is that we created the `--mode dev-container` that bundled well-known config properties to improve the startup time for dev containers:

```
--mode dev-container
    Bundled flags:
        * --overprovisioned
        * --reserve-memory 0M
        * --check=false
        * --unsafe-bypass-fsync
    Bundled cluster properties:
        * auto_create_topics_enabled: true
        * group_topic_partitions: 3
        * storage_min_free_bytes: 10485760 (10MiB)
        * topic_partitions_per_shard: 1000
        * fetch_reads_debounce_timeout: 10
        * group_initial_rebalance_delay: 0
        * log_segment_size_min: 1
```

During the implementation we discussed how this new dev-container mode played with the `developer_mode: true/false` and we decided to go with:
- If developer_mode is enabled, then use mode dev-container.
- If developer_mode is disabled, don't change anything and leave the defaults.

**References:**
- Original Discussion: https://github.com/redpanda-data/redpanda/discussions/2122
- Original Issue: https://github.com/redpanda-data/redpanda/issues/5208
- PR with some discussion about the properties: https://github.com/redpanda-data/redpanda/pull/5628

micheleRP commented:
Thanks @r-vasquez! Can you please confirm that users who want to deploy Redpanda in a dev/test environment should still follow these instructions, which is similar to the instructions to deploy for production, just not setting `rpk redpanda mode production`. For more info, should we link to [development mode](http://localhost:3000/docs/reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode) or [--mode dev container](https://docs.redpanda.com/docs/reference/rpk/rpk-redpanda/rpk-redpanda-start/#setting-up-a-mode). And can you also please confirm if any of [these hardware/software prereqs](https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment/) are required here?

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,185 @@
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+You can deploy Redpanda using well-known configuration properties for development or test environments. For example, in [development mode](../../../../../../reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode), the default `group_topic_partitions` is 3, and the default `storage_min_free_bytes` is 1 G. In production mode, the default `group_topic_partitions` is 16, and the default `stor ...
```
'optimized for development or test environments.  These configuration settings may use fewer system resources and simplify setup, but are not suitable for production deployments. Please see <link> for production deployment instructions.'

micheleRP replied:
Thanks, @mattschumpert! Please see the updated draft. Should I also remove the examples (For example, in development mode, the default `group_topic_partitions` is 3, and the default `storage_min_free_bytes` is 1 G. In production mode, the default `group_topic_partitions` is 16, and the default `storage_min_free_bytes` is 5 G.)?

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,185 @@
+You can deploy Redpanda using well-known configuration properties for development or test environments. For example, in [development mode](../../../../../../reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode), the default `group_topic_partitions` is 3, and the default `storage_min_free_bytes` is 1 G. In production mode, the default `group_topic_partitions` is 16, and the default `stor ...
+
+:::note
+- Development mode bypasses `fsync`, which results in unrealistically fast clusters and potential data loss.
```
'Development mode bypasses 'fsync', acknowledging messages before they are stored to disk. This reduces the durability of messages, may give unrealistic performance expectations with respect to a production clusters, and could potentially lead to data loss if used in a production configuration'

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,185 @@
+Follow these requirements and recommendations:
+
+- Operating system
+  - Minimum required version of the Linux kernel: <Versions name="linux-kernel"/>
```
In the preview I dont see any version. This can be found in our supported platforms docs.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,185 @@
+
+## Install Redpanda Console
+
+Redpanda Console is a developer-friendly web UI for managing and debugging your Redpanda cluster and your applications.
```
We recommend always installing Redpanda Console alongside a Redpanda cluster.

larsenpanda reviewed (approved):
This looks good to me

micheleRP commented:
> Thanks @r-vasquez! Can you please confirm that users who want to deploy Redpanda in a dev/test environment should still follow these instructions, which is similar to the instructions to deploy for production, just not setting `rpk redpanda mode production`. For more info, should we link to [development mode](http://localhost:3000/docs/reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode) or [--mode dev container](https://docs.redpanda.com/docs/reference/rpk/rpk-redpanda/rpk-redpanda-start/#setting-up-a-mode). And can you also please confirm if any of [these hardware/software prereqs](https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment/) are required here?

@r-vasquez or @andrwng: We need someone from engineering to approve this PR. For more info, should we link to [development mode](http://localhost:3000/docs/reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode) or [--mode dev container](https://docs.redpanda.com/docs/reference/rpk/rpk-redpanda/rpk-redpanda-start/#setting-up-a-mode)? And can you please confirm if we should keep or remove [these hardware/software prereqs](https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment/)?

andrwng commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,187 @@
+import TabItem from '@theme/TabItem';
+import Versions from '../../../../../shared/_versions.mdx'
+
+You can deploy Redpanda using well-known configuration properties optimized for a development or test environment. This configuration uses less system resources and simplifies setup, but it's not suitable for a production deployment. For example, in [development mode](../../../../../../reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode), the default `group_topic_partitions` is 3, and  ...
```
nit: GiB instead of G

andrwng commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
````diff
@@ -0,0 +1,187 @@
+
+```bash
+sudo rpk redpanda config bootstrap --self <private-ip> --ips <seed-node-ips> && \
+sudo rpk redpanda config set redpanda.empty_seed_starts_cluster false && \
````
This is missing a line? I think

sudo systemctl start redpanda

micheleRP replied:
Thanks, Andrew! I've updated the doc.

andrwng commented:
>should keep or remove [these hardware/software prereqs](https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment/)?

Everything recommended for production would still be _recommended_ for a development environment. OS and other requirements are definitely still applicable.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,188 @@
+When a Redpanda cluster starts, it instantiates a controller Raft group with all the seed nodes that are specified in the `--ips` flag. After all seed nodes complete their startup procedure and become accessible, the cluster is then available. After that, non-seed nodes start up and are added to the cluster.
+
+:::info
+- Redpanda strongly recommends at least three seed nodes when forming a cluster. A larger number of seed nodes increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data.
```
I wonder if it makes sense to include this in the 'dev mode' docs when you might be running just a single node cluster in dev mode.  This robustness against multiple node failures causing fracture is really only needed for production.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,188 @@
+:::info
+- Redpanda strongly recommends at least three seed nodes when forming a cluster. A larger number of seed nodes increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data.
+- It's important to have one or more seed nodes in each fault domain (such as rack or cloud AZ). A higher number provides a stronger guarantee that clusters don’t fracture unintentionally.
+- It's possible to change the seed nodes for a short period of time after a cluster has been created. For example, you may want to designate one additional broker as a seed node to increase availability. To do this without cluster downtime, add the new broker to [`seed_servers`](../../../../../../reference/node-properties/) and restart Redpanda to apply the change on a broker-by-broker basis.
```
This is kind of ambiguous.  @mmaslankaprv  can we give any more specific guidance here?

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
````diff
@@ -0,0 +1,188 @@
+To verify that the Redpanda cluster is up and running, check the logs:
+
+```bash
+journalctl -u redpanda
````
should we not rather advise using rpk to check broker status to verify we're fully up and running @twmb ?

twmb replied:
Sure, we can say to run `rpk cluster info` to detect if rpk can locally connect and print some information

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
````diff
@@ -0,0 +1,188 @@
+journalctl -u redpanda
+```
+
+If topics were initially created in a test environment with a replication factor of `1`, use `rpk topic alter-config` to change the topic replication factor: 
````
@micheleRP  this is really dev --> prod advice we should skip here.  If you're starting out in dev and created topics with replication factor 1 you'd likely fix then only when going prod.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,188 @@
+
+## Next steps
+
+If clients will connect from a different subnet, see [Configuring Listeners](../../../../../../manage/security/listener-configuration).
```
@micheleRP  connectivity is a good one to hit on in dev clusters. any other advice we have on connectivity in addition to this is helpful here.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,188 @@
+
+- [Redpanda Cluster Configuration](../../../../../../manage/cluster-maintenance/configuration)
+- [Redpanda Console Configuration](../../../../../../reference/console/config)
+- [Work with Schema Registry](../../../../../../manage/schema-registry)
```
@micheleRP  what about Pandaproxy? https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/develop/http-proxy/

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/production/index.mdx:
```diff
@@ -8,10 +8,14 @@ title: Production Deployment
-  
-    You can use `rpk` to set up a production cluster of Redpanda using our default deployment or by writing a custom deployment script.
+
+   Deploy Redpanda for production with a default deployment, which uses recommended deployment tools, or with a custom deployment, which uses unsupported deployment tools.
```
'unsupported deployment tools' is a bit misleading and harsh.

We support custom deployments, we just strongly recommend using our automation tooling. We also make the point in the production deployment page, not need to make it here.  This page is for navigation only.

I'd reorder these options to put the recommended 'Automate Deployment for Production (recommended)' first (as the 2nd option), and 'Deploy Redpanda for production with a default deployment or custom automation' as the 3rd option in the list and leave it at that

micheleRP replied:
thanks, I've updated the draft.

mattschumpert commented on docs/deploy/deployment-option/self-hosted/manual/index.mdx:
@rkruze  should we be calling this section 'Manual' or just 'Bare Metal'?  'Manual' feels quite odd when in it we're advising using our deployment automation tooling as the recommended option.  The hierarchy is Self hosted --> K8s or [blank].  I would think 'bare metal' or similar:

https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/

micheleRP replied:
> @rkruze should we be calling this section 'Manual' or just 'Bare Metal'? 'Manual' feels quite odd when in it we're advising using our deployment automation tooling as the recommended option. The hierarchy is Self hosted --> K8s or [blank]. I would think 'bare metal' or similar:
> 
> https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/

@Feediver1 & @JakeSCahill please also comment on this. Our pages differentiate with K8s vs Linux, like [here ](https://docs.redpanda.com/docs/manage/tiered-storage/) or [here](https://docs.redpanda.com/docs/manage/monitoring/).

micheleRP replied:
@mattschumpert Doc team agreed we should rename to Manual to Linux. I've updated the sidebar and index pages: https://deploy-preview-1442--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/

Feediver1 commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,183 @@
+
+You can deploy Redpanda using well-known configuration properties optimized for a development or test environment. This configuration uses less system resources and simplifies setup, but it's not suitable for a production deployment. For example, in [development mode](../../../../../../reference/rpk/rpk-redpanda/rpk-redpanda-mode/#development-mode), the default `group_topic_partitions` is 3, and  ...
+
+:::note
```
This seems important--no?

micheleRP replied:
I'll change to important

Feediver1 commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,183 @@
+
+## Verify the installation
+
+To verify that the Redpanda cluster is up and running, check the logs:
```
Seems like if we are going to tell users to check the logs, we should show what the results are they should be looking for, no?

micheleRP replied:
This step changed with a previous review comment. Verify now by running `rpk cluster info`.

Feediver1 commented on docs/deploy/deployment-option/self-hosted/manual/production/dev-deployment.mdx:
```diff
@@ -0,0 +1,183 @@
+
+## Next steps
+
+If clients will connect from a different subnet, see [Configuring Listeners](../../../../../../manage/security/listener-configuration).
```
```suggestion
If clients connect from a different subnet, see [Configuring Listeners](../../../../../../manage/security/listener-configuration).
```

Feediver1 reviewed (approved):
See questions
